File: packages/nest/src/changes.controller.ts
import {
  BadRequestException,
  Controller,
  ForbiddenException,
  Get,
  Inject,
  NotImplementedException,
  Query,
  Req,
} from '@nestjs/common';
import type { ChangesResponse, Cursor } from '@maayo/protocol';
import { MAAYO_OPTIONS } from './maayo.constants';
import type { MaayoModuleOptions } from './maayo.options';
import type { SavedMutation } from './interfaces';

@Controller('sync')
export class ChangesController {
  constructor(@Inject(MAAYO_OPTIONS) private readonly options: MaayoModuleOptions) {}

  @Get('changes')
  async pull(
    @Query('channel') channel: string,
    @Query('since') since?: string,
    @Query('lastMutationId') lastMutationId?: string,
    @Query('limit') limitStr?: string,
    @Req() req: unknown = undefined,
  ): Promise<ChangesResponse> {
    const { store, authorizer, defaultLimit = 500 } = this.options;

    if (authorizer && !(await authorizer.canPull(req, channel))) {
      throw new ForbiddenException(`unauthorized for channel ${channel}`);
    }

    const limit = clamp(parseInt(limitStr ?? String(defaultLimit), 10) || defaultLimit, 1, 2000);
    const hasSince = typeof since === 'string' && since.trim().length > 0;
    const hasLastMutationId = typeof lastMutationId === 'string' && lastMutationId.trim().length > 0;
    if (hasSince !== hasLastMutationId) {
      throw new BadRequestException('since and lastMutationId must be provided together');
    }

    let rows: SavedMutation[];
    if (hasSince && hasLastMutationId) {
      const sinceDate = new Date(since);
      if (Number.isNaN(sinceDate.getTime())) {
        throw new BadRequestException('since must be a valid ISO-8601 timestamp');
      }
      if (!store.findChangesByCursor) {
        throw new NotImplementedException('store does not support compound-cursor pagination');
      }
      rows = await store.findChangesByCursor(channel, sinceDate, lastMutationId, limit + 1);
    } else {
      rows = await store.findChanges(channel, null, limit + 1);
    }

    const hasMore = rows.length > limit;
    const page = hasMore ? rows.slice(0, limit) : rows;

    return {
      channel,
      mutations: page.map((r) => r.mutation),
      cursor: buildCursor(page),
      hasMore,
    };
  }
}

function buildCursor(page: SavedMutation[]): Cursor {
  const last = page[page.length - 1];
  if (!last) return { lastMutationId: null, lastReceivedAt: null };
  return {
    lastMutationId: last.mutation.id,
    lastReceivedAt: last.receivedAt.toISOString(),
  };
}

function clamp(n: number, min: number, max: number): number {
  return Math.min(Math.max(n, min), max);
}

File: packages/spring/src/main/kotlin/dev/maayo/spring/api/ChangesController.kt
package dev.maayo.spring.api

import dev.maayo.spring.ChannelAuthorizer
import dev.maayo.spring.MaayoProperties
import dev.maayo.spring.MaayoRepository
import dev.maayo.spring.SavedMutation
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException
import java.security.Principal
import java.time.Instant

@RestController
@RequestMapping("/sync")
class ChangesController(
    private val repository: MaayoRepository,
    private val authorizer: ChannelAuthorizer,
    private val properties: MaayoProperties,
) {
    @GetMapping("/changes")
    fun pull(
        @RequestParam channel: String,
        @RequestParam(required = false) since: String?,
        @RequestParam(required = false) lastMutationId: String?,
        @RequestParam(required = false) limit: Int?,
        principal: Principal?,
    ): ChangesResponse {
        if (!authorizer.canPull(principal, channel)) {
            throw ResponseStatusException(HttpStatus.FORBIDDEN, "unauthorized for channel $channel")
        }

        val effectiveLimit = (limit ?: properties.defaultLimit).coerceIn(1, 2000)
        val sinceInstant = since?.let { runCatching { Instant.parse(it) }.getOrNull() }
        val rows = repository.findChanges(channel, sinceInstant, lastMutationId, effectiveLimit + 1)

        val hasMore = rows.size > effectiveLimit
        val page = if (hasMore) rows.dropLast(1) else rows

        val cursor = buildCursor(page)

        return ChangesResponse(
            channel = channel,
            mutations = page.map { it.toDto() },
            cursor = cursor,
            hasMore = hasMore,
        )
    }

    private fun buildCursor(page: List<SavedMutation>): Cursor {
        val last = page.lastOrNull() ?: return Cursor(null, null)
        return Cursor(
            lastMutationId = last.mutation.id,
            lastReceivedAt = last.receivedAt.toString(),
        )
    }

    private fun SavedMutation.toDto() = mutation
}
